### ISSUE TITLE
Possible overhaul of instructions.

### ISSUE THREAD

**Issue by erp12** [erp12/pyshgp#85]
> Note: Maybe this whole issue can be ignored if we want to migrate the push interpreter to [CPushPush](https://github.com/Killeroid/CPushPush).

## Instructions 

We should easily be able to rely more on inheritance and add more functionality to the constructor of `Instruction` in order to remove a lot of code duplication.

See the way instructions are made in Propel and CPushPush for examples of patterns that are much better than what is currently in pyshgp.

## Instruction Set

Each python module which defines Push Instructions should define `__all__` and then we can safely use `import *` in the instructions `__init__.py`. This should make the whole instructions sub-package much easier to understand.

## Tests
This time we should put a lot more thought into removing the code duplication. Here is a sketch of what I would currently suggest, although more thought should maybe be put into this:

```py
common_tests = [
    [
        {'_integer': [1, 2, 3]},
        {'_integer': [1, 2]},
        '_integer_pop'
    ],
    [
        {'_boolean': [True]},
        {'_boolean': []},
        '_boolean_pop'
    ],
    [
        {'_string': ['A']},
        {'_string': ['A', 'A']},
        '_boolean_pop'
    ],
    [
        {'_float': [1.5, 2.3]},
        {'_float': [2.3, 1.5]},
        '_float_swap'
    ],
    [
        {'_integer': [1, 2, 3]},
        {'_integer': [2, 3, 1]},
        '_integer_rot'
    ]
]

for test in common_tests:
    assert run_test(*test)
```

**Comment by erp12** [comment]
5394078912dada49ee31645cb725024e24fbcb1d is a minor refactor that will make the adding of "docstrings" to instructions easier.

An initial pass as restructuring the Instruction object so that the `func` attribute wouldn't have to handle to popping of arguments and pushing of results was done as part of this work, but it was quickly scrapped.

The reason I am holding off on this change is because instructions need to remain atomic. If the popping of instruction arguments is abstracted away, it will have to revert in the event that some other condition causes the instruction to not run. (ie. If the denominator of division is 0)

**Comment by erp12** [comment]
Fixed with #102
